Source/DafnyLanguageServer/Workspace/DocumentDatabase.cs

```diff
@@ -99,65 +109,107 @@ ISymbolTableRelocator symbolTableRelocator
       }
       databaseEntry.CancelPendingUpdates();
       var cancellationSource = new CancellationTokenSource();
+      var previousDocumentTask = FirstSuccessfulTask(databaseEntry.VerifiedDocument, databaseEntry.ResolvedDocument);
+      var resolvedDocumentTask = ApplyChangesAsync(previousDocumentTask, documentChange, cancellationSource.Token);
+      var verifiedDocument = VerifyAsync(resolvedDocumentTask, VerifyOnChange, cancellationSource.Token);
       var updatedEntry = new DocumentEntry(
         documentChange.TextDocument.Version,
-        Task.Run(async () => await ApplyChangesAsync(await databaseEntry.Document, documentChange, cancellationSource.Token)),
+        resolvedDocumentTask,
+        verifiedDocument,
         cancellationSource
       );
       documents[documentUri] = updatedEntry;
-      return await updatedEntry.Document;
+      return resolvedDocumentTask.ToObservable().Concat(verifiedDocument.ToObservable());
+    }
+
+
+#pragma warning disable VSTHRD200
+    private static Task<T> FirstSuccessfulTask<T>(params Task<T>[] tasks) {
+#pragma warning restore VSTHRD200
+      var taskList = tasks.ToList();
+      var tcs = new TaskCompletionSource<T>();
+      int remainingTasks = taskList.Count;
+      foreach (var task in taskList) {
+#pragma warning disable VSTHRD110
+        task.ContinueWith(t => {
+#pragma warning restore VSTHRD110
+          if (task.Status == TaskStatus.RanToCompletion) {
+            tcs.TrySetResult(t.Result);
+          } else if (Interlocked.Decrement(ref remainingTasks) == 0) {
+            if (tasks.Any(otherTask => otherTask.IsCanceled)) {
+              tcs.SetCanceled();
+            } else {
+              tcs.SetException(new AggregateException(tasks.SelectMany(t1 => t1.Exception!.InnerExceptions)));
+            }
+          }
+        }, TaskScheduler.Default);
+      }
+      return tcs.Task;
     }
 
-    private async Task<DafnyDocument> ApplyChangesAsync(DafnyDocument oldDocument, DidChangeTextDocumentParams documentChange, CancellationToken cancellationToken) {
+    private async Task<DafnyDocument> ApplyChangesAsync(Task<DafnyDocument> oldDocumentTask, DidChangeTextDocumentParams documentChange, CancellationToken cancellationToken) {
+#pragma warning disable VSTHRD003
```

Comment:
Could you just await the document at the call site to get rid of this warning? That way you also help reduce the number of created tasks.

Reply:
I can't because it's not an async method.